### PR TITLE
Pull requests support in the CI

### PR DESCRIPTION
**Pull request by depau**
- [x] Support GitHub pull requests
- [x] Support GitLab merge requests
- GitHub/Lab API calls to report CI status
  - [x] GitHub Checks app implementation
  - [x] GitLab PR comments

### CODE REVIEW

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: A bit sad. Do you envision an opportunity to unify these?

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]:
```suggestion
```

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: What's going on here? Doc

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Unused arguments are a bit sketchy to me. We could introduce them when needed.
But no biggie.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Do you actually expect all of them?

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Consider logging

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: No need to check this in the previous step?

File: .orchestra/ci/ci-hook/environment.example
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/environment.example]: * Missing new line at end of file
* What is this for?

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: NOTETOSELF: review

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: We can indeed remove all manually-added webhooks and use the app only. Should I go ahead?

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: The workflow for GitHub is:

1. User pushes a commit to a managed repo or to a PR targeting the repo
2. GitHub sends a `check_suite` webhook with the `requested` action
3. ---> We create a new check run (and, implicitly, a check suite) **[this is what happens in this code]**
4. GitHub sends a `check_run` webhook with action `created`
5. We actually start the CI

In step 2 we additionally handle the `rerequested` case (when the job failed and the user wants to re-run it), where we still need to create a new check run. The process is the same but the webhook payload we receive is slightly different, hence the `"x" in data and data["x"] ...`

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: The GH API does differentiate between "success causes" by HTTP status code in some cases. 200 to 202 are all good for the purposes of this code. I didn't bother checking the exact cases, someone else might care about the difference between "ok", "resource created", "resource updated", but our application doesn't.

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: Could make sense, but it is also useful to return it as a response since you can visit the GitHub app page, view failed webhooks and see the response.

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: No. `requested` action is sent to specific applications, `created` is broadcast.

File: .orchestra/ci/ci-hook/environment.example
Comment (depau) [reply .orchestra/ci/ci-hook/environment.example]: - `GITHUB_PRIVATE_KEY_BASE64` → auth to the GitHub API
- `GITHUB_APP_SECRET` → hmac shared secret for webhook authentication

ack for `\n`.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Yes, let's do this.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: OK, add doc about this.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Sure, let's do both

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: Done

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: Done (old webhooks are still configured but they aren't currently being accepted by the deployed hook).

I entirely removed the `push` hook since we still get a check run request for pushes to the configured repos, in order to display the checkmarks

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: Done

File: .orchestra/ci/ci-run.sh
Comment (aleclearmind) [inline .orchestra/ci/ci-run.sh]: Comment what's going on here, it's very cryptic otherwise.

Comment (depau) [reply .orchestra/ci/ci-run.sh]: Done

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: * I'm really really suspicious of large `try` blocks. Make it as small as possible.
* Also, I'd prefer an hard failure here. If this happens, it means we need to fix something.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Log to `stderr`

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Is this actually pull-request specific?

Can you recap for me what webhooks are we listening to?
I understand we're no longer listening to push webhooks, in favor of checks webhooks. But AFAIU those are not pull-request specific, so maybe we can have a more appropriate name?

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: So we no longer look at CI variables? That's kinda sad, sometimes I used to run the CI by hand with a custom value for `TARGET_COMPONENTS`. Now to do the same I need to log on the server, edit the config and restart the daemon.
Is that right?

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: 1. Does this contain any sensitive information?
2. This does not affect the behavior of the job, it's just metadata for later consumption. Is this right?

Add some docs.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Why do this only if `username not in allowed_to_push`?

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Make a `log` function and use `stderr`.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]: Find a more suitable variable name

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [inline .orchestra/ci/ci-hook/app.py]:
```suggestion
@app.route('/ci-hook/gitlab-ci', methods=["POST"])
```

Can you rename the method and URL to be more explicit on the fact that this handles the result of a CI job/pipeline?
I was initially confused and thought we no longer triggered a pipeline upon a push on gitlab.

File: .orchestra/ci/ci-run.sh
Comment (aleclearmind) [inline .orchestra/ci/ci-run.sh]: This looks terrible. Can we have a `bash` function with a decent name and hiding the complexity of the `bash` syntax?

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Should I remove all the webhooks from all repositories?

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: GH:
- pull_request:
  - opened
  - synchronize (push)
- check_run/check_suite:
  - requested (push to the main repo)
  - rerequested (user clicks re-run)

GL:
- Push
- Merge Request

> so maybe we can have a more appropriate name?

I'd say just `user_options`, it's used always except when pushing

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: If you run it by hand on the gitlab UI it uses your value. This overrides the default when triggered by the CI hook.

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: 1. No
2. Yes, it only contains data required to submit back the job status to the pull request. If not set, no status updates are sent.

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: The full condition is `username in allowed_to_push and base_repo_url == repo_url` and it's important.

If it is true it means that you're triggering the CI from **a push to the repo itself** (not a fork) and that you're trying to promote a branch. So the SSH user_options should be used.

If it is false it means that it is either being launched from a fork or from another user. In this case we want to add the user namespace to the options, and we want internal users to be able to access the private repos but by using the internal impersonation token authentication, not the SSH keys with push access.

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: On GH, yes.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Gotcha, add some docs.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Gotcha, docs docs docs.

File: .orchestra/ci/ci-hook/app.py
Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Done, I've also removed the webhook for revng-c on gitlab.

Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: > Done, I've also removed the webhook for revng-c on gitlab.

Do you no longer want it to trigger the CI? On GitLab it is still needed, you need "push" and "merge request" for all repos.

Comment (aleclearmind) [reply .orchestra/ci/ci-hook/app.py]: Ah shit. Can you set it up again? `:P`

**Comment by depau** [comment]
@aleclearmind since I made lots of small changes, I added them to a single commit which I do not expect you to merge. If everything looks good to you LMK and I'll `rebase -i` it properly before merging.

All the comments with :rocket: have been addressed on my side.

File: .orchestra/ci/ci-hook/app.py
Comment (depau) [reply .orchestra/ci/ci-hook/app.py]: Done

**Comment by aleclearmind** [comment]
LGTM, please squash last commit as appropriate so we can proceed.

**Comment by depau** [comment]
@aleclearmind done

**Pull request closed without merging**
